### PR TITLE
Add union to test_all_types, and arrow and json R/W

### PR DESCRIPTION
**Pull request by Mause**
This started out adding support to the Python `.description` field for the union data type, but I realised that we didn't have it in the `test_all_types` table function.

 * Added support for R/W in Arrow, specifically as a Sparse Union. Reading Dense Unions will still fail
 * Added read support for Unions in JSON. Write support was already there, so I just followed the existing pattern
 * Corrected a number of locations where the Union tag type was incorrect - a `uint8_t` instead of a `int8_t`
 * Parquet doesn't support unions (see https://issues.apache.org/jira/browse/PARQUET-756)
 * Added support for reading unions into numpy/pandas (as object type)
 * R support was left on the table, I'm not familiar enough with R to do this without guidance
 * Basic Java union read support is there, but only stringly typed - I'll add full support when I added Struct support (soon)

### CODE REVIEW

File: extension/json/json_functions/json_transform.cpp [hunk 1]
```diff
@@ -689,6 +689,55 @@ bool TransformToJSON(yyjson_val *vals[], yyjson_alc *alc, Vector &result, const
...
+		if (!yyjson_is_obj(vals[i])) {
+			if (options.error_missing_key) {
+				throw InvalidInputException("Expected an object representing a union");
+			} else {
+				result.SetValue(i, Value(nullptr));
+				continue;
+			}
+		}
+
+		yyjson_val *key, *val;
```
Comment (lnkuiper) [inline extension/json/json_functions/json_transform.cpp:712]: I would recommend using `yyjson_obj_foreach` here, it makes the loop more readable. There's a few examples in this file.

File: extension/json/json_functions/json_transform.cpp [hunk 2]
```diff
@@ -689,6 +689,55 @@ bool TransformToJSON(yyjson_val *vals[], yyjson_alc *alc, Vector &result, const
...
+		yyjson_obj_iter iter;
+		yyjson_obj_iter_init(vals[i], &iter);
+		key = yyjson_obj_iter_next(&iter);
+		if (key == nullptr) {
```
Comment (lnkuiper) [inline extension/json/json_functions/json_transform.cpp:716]: `key` should not be able to ever be `nullptr` when using `yyjson_obj_foreach`. Perhaps it's worth checking `yyjson_get_len` here to check if it's an empty object before going into the loop

File: extension/json/json_functions/json_transform.cpp [hunk 3]
```diff
@@ -689,6 +689,55 @@ bool TransformToJSON(yyjson_val *vals[], yyjson_alc *alc, Vector &result, const
...
+			// empty object?
+			result.SetValue(i, Value(nullptr));
```
Comment (lnkuiper) [inline extension/json/json_functions/json_transform.cpp:718]: For setting results to `NULL`, we use `Validity`
```c++
auto &validity = FlatVector::Validity(result);
result.SetInvalid(i);
```

File: extension/json/json_functions/json_transform.cpp [hunk 4]
```diff
@@ -689,6 +689,55 @@ bool TransformToJSON(yyjson_val *vals[], yyjson_alc *alc, Vector &result, const
...
+			continue;
+		}
+
+		val = yyjson_obj_iter_get_val(key);
+
+		auto tag = std::find(names.begin(), names.end(), unsafe_yyjson_get_str(key));
+		if (tag == names.end()) {
+			throw InvalidInputException("Found object containing unknown key instead of union");
```
Comment (lnkuiper) [inline extension/json/json_functions/json_transform.cpp:726]: Same as before, could you use the new error handling pattern?

File: extension/json/json_functions/json_transform.cpp [hunk 5]
```diff
@@ -689,6 +689,55 @@ bool TransformToJSON(yyjson_val *vals[], yyjson_alc *alc, Vector &result, const
...
+		idx_t actualtag = tag - names.begin();
+
+		Vector single(UnionType::GetMemberType(type, actualtag), 1);
+		// should we use this return value
+		JSONTransform::Transform(&val, alc, single, 1, options);
+
+		result.SetValue(i, Value::UNION(fields, actualtag, single.GetValue(0)));
+	}
+
+	return true;
```
Comment (lnkuiper) [inline extension/json/json_functions/json_transform.cpp:738]: This should return `success` - whether we were able to transform all values. So this answers your question: Yes! We should use the return value :)

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Thanks for the PR! I have left some comments about the `JSONTransform` implementation.

I'm also wondering about how other systems map from JSON to UNION. Do you have any idea? Perhaps we're the only system out there that does this. I always thought that reading JSON like this:
```json
{"value": 42}
{"value": [1, 2, 3]}
```
Could be read using `UNION(int INTEGER, int_list INT[])` without having to wrap them in an object with the keys "int" or "int_list".

But now that I see your implementation, I'm not sure what makes the most sense, especially since I'm not a heavy JSON user myself.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great.

As this is adding unions to the arrow conversion - could we add some union-specific arrow conversion tests in the Python client? In particular nested unions, unions with lists/structs inside them, unions inside lists/structs, all are not covered yet but could potentially cause problems.

**Review by Maxxen — APPROVED** [review]

**Comment by Mytherin** [comment]
Thanks for the fixes! Could you just have a look at the merge conflict? The CI failure in `test/sql/json/test_json_copy.test_slow` also seems related to the changes made in this PR.

**Comment by Mytherin** [comment]
Could you have a look at the failing CI?

**Comment by Mause** [comment]
Failures don't look related my changes?

**Comment by Mytherin** [comment]
Could you merge with master?

**Pull request merged**